File: src/yazses/stt/filters/disfluency.py
"""Offline disfluency filter — removes filler words, repetitions, and self-corrections."""
from __future__ import annotations

import re
from dataclasses import dataclass

from yazses.config import DisfluencyConfig


@dataclass
class FilterResult:
    text: str
    chars_removed: int


def filter_transcript(text: str, config: DisfluencyConfig | None = None) -> FilterResult:
    """Apply three-pass disfluency filter. Returns cleaned text and chars removed.

    Rule A: filler word removal (case-insensitive word boundary regex).
            Guard: do NOT remove tokens that contain uppercase, underscore, slash, or dot
            (those are proper nouns or code identifiers). Exception: a capitalised filler
            at utterance position 0 is removed only when it is an unambiguous filler —
            Whisper capitalises sentence starts, so leading "Um"/"Uh" would otherwise
            never be stripped (see #117). Content words like "Right"/"Like" and
            multi-word fillers like "You know" stay protected (see #120).
    Rule B: consecutive 2-gram deduplication (repeat until stable).
    Rule B.5: sub-word repetition & prolongation collapse (opt-in; ADR-015).
    Rule C: self-correction trigger detection — if found, remove from last sentence
            boundary before the trigger, through to end of trigger phrase.
    """
    if config is None:
        config = DisfluencyConfig()
    if not config.enabled or not text.strip():
        return FilterResult(text=text, chars_removed=0)

    original_len = len(text)

    # Rule A — filler word removal
    text = _remove_fillers(text, config.filler_words)

    # Rule B — 2-gram consecutive dedup
    text = _dedup_2grams(text)

    # Rule B.5 — sub-word repetition & prolongation collapse (opt-in; ADR-015)
    text = _collapse_dysfluencies(text, config)

    # Rule C — self-correction rollback
    text = _apply_self_corrections(text, config.self_correction_triggers)

    # Normalise multiple spaces
    text = re.sub(r'  +', ' ', text).strip()

    return FilterResult(text=text, chars_removed=max(0, original_len - len(text)))


def _is_protected(token: str) -> bool:
    """Return True if this token should NOT be altered (proper noun or code id)."""
    return bool(
        any(c.isupper() for c in token)
        or '_' in token
        or '/' in token
        or '.' in token
    )


# Sentence punctuation that Whisper hangs off a hesitation ("Uh...", "Um,").
# It is not evidence of a code identifier, so it is trimmed before that check.
_TRAILING_PUNCTUATION = '.,;:!?…'


def _is_code_or_path_token(token: str) -> bool:
    """True when the token looks like a code identifier or path/URL fragment.

    Only a dot *inside* the token counts. Whisper writes a hesitation as
    ``"Uh..."``, and treating that trailing dot like the one in ``main.py`` left
    a leading filler permanently protected (issue #122). ``Actually.`` at the
    end of a sentence stays safe regardless: it is not a hesitation particle,
    so ``_is_unambiguous_filler`` refuses it before this check matters.
    """
    core = token.rstrip(_TRAILING_PUNCTUATION)
    return '_' in core or '/' in core or '.' in core


# Non-lexical hesitation sounds. These are never ordinary English words, which
# is what makes them safe to strip at utterance position 0 despite Whisper's
# capitalisation.
#
# The membership test is lexical, not phonetic: a particle qualifies only when
# it has no dictionary sense outside interjection. "err" looked like it belonged
# here and does not — it is a verb ("to err is human"), so it was removed from
# both this set and the default filler list (issue #125). "ah" stays: it is an
# interjection in every sense, so dropping it costs tone, never meaning.
_HESITATION_PARTICLES = frozenset({'um', 'uh', 'er', 'ah', 'hmm'})


def _is_unambiguous_filler(filler: str) -> bool:
    """True when ``filler`` contains a non-lexical hesitation particle.

    Position 0 is the one place Whisper's capitalisation is uninformative, so
    the uppercase guard is relaxed there — but only for fillers that cannot
    open a real sentence. A filler qualifies when any of its words is a
    hesitation particle, which admits ``"so um"`` / ``"so uh"`` (issue #122)
    while still refusing ``"you know"``, ``"i mean"`` and ``"okay so"``: those
    are three real words that legitimately start a sentence, and deleting one
    is worse than leaving a filler in (issue #120).
    """
    return any(word in _HESITATION_PARTICLES for word in filler.split())


def _is_utterance_initial(text: str, match_start: int) -> bool:
    """True when ``match_start`` is the first non-whitespace content in ``text``."""
    return not text[:match_start].strip()


def _collapse_prolongations(text: str, min_run: int) -> str:
    """Collapse a run of the same letter of length >= ``min_run`` to one letter.

    Per token; protected tokens (proper nouns / code / URLs) are left untouched.
    English double letters (run length 2) stay safe when ``min_run`` >= 3.
    """
    if min_run < 2:
        return text
    pattern = re.compile(rf'([a-zA-Z])\1{{{min_run - 1},}}')
    out: list[str] = []
    for token in text.split():
        if _is_protected(token):
            out.append(token)
        else:
            out.append(pattern.sub(r'\1', token))
    return ' '.join(out)


def _collapse_repetitions(text: str, max_fragment_len: int) -> str:
    """Collapse stutter repetitions: hyphenated false starts, short fragment
    runs, and heavy unigram repeats. Conservative by design — intentional
    hyphenation (re-read), emphasis pairs (very very), and protected tokens
    are left untouched."""

    def _fix_hyphen(tok: str) -> str:
        if '-' not in tok or _is_protected(tok):
            return tok
        parts = tok.split('-')
        if len(parts) >= 3:  # >= 2 leading fragments + final word
            lead, final = parts[:-1], parts[-1]
            frag = lead[0]
            if final and frag.isalpha() and all(p == frag for p in lead) \
               and final.lower().startswith(frag.lower()):
                return final
        return tok

    tokens = [_fix_hyphen(t) for t in text.split()]
    result: list[str] = []
    i, n = 0, len(tokens)
    while i < n:
        tok = tokens[i]
        # (b) >=2 identical short fragments followed by a longer word they prefix
        if tok.isalpha() and not _is_protected(tok) and len(tok) <= max_fragment_len:
            j = i
            while j < n and tokens[j] == tok:
                j += 1
            if (j - i) >= 2 and j < n and not _is_protected(tokens[j]) \
               and len(tokens[j]) > len(tok) and tokens[j].lower().startswith(tok.lower()):
                result.append(tokens[j])
                i = j + 1
                continue
        # (c) unigram run of length >= 3
        if not _is_protected(tok):
            j = i
            while j < n and tokens[j] == tok:
                j += 1
            if (j - i) >= 3:
                result.append(tok)
                i = j
                continue
        result.append(tok)
        i += 1
    return ' '.join(result)


def _collapse_dysfluencies(text: str, config: DisfluencyConfig) -> str:
    """Opt-in Rule B.5 — sub-word repetition + prolongation collapse (ADR-015)."""
    if config.collapse_repetitions:
        text = _collapse_repetitions(text, config.repetition_max_fragment_len)
    if config.collapse_prolongations:
        text = _collapse_prolongations(text, config.prolongation_min_run)
    return text


def _hyphen_parts(token: str) -> list[str]:
    """The hyphen-separated parts of a token, ignoring surrounding punctuation."""
    core = token.strip(_TRAILING_PUNCTUATION + '"\'()[]')
    return [p for p in core.split('-') if p]


def _is_all_filler_hyphenated(token: str, fillers: frozenset[str]) -> bool:
    """True when every part of a hyphenated token is a filler word.

    ``um-um`` and ``uh-uh`` qualify; ``a-a-actually`` does not, because
    ``actually`` is a real word the user said.
    """
    parts = _hyphen_parts(token)
    return len(parts) > 1 and all(p.lower() in fillers for p in parts)


def _drop_all_filler_hyphen_tokens(text: str, fillers: frozenset[str]) -> str:
    """Remove hyphenated tokens that are nothing but fillers, whole.

    Removing them part-by-part is what left an orphan ``-`` glued to the next
    word (``um-um the meeting`` -> ``-the meeting``).
    """
    kept = [t for t in text.split(' ') if not _is_all_filler_hyphenated(t, fillers)]
    return ' '.join(kept)


def _tidy_orphaned_punctuation(text: str) -> str:
    """Repair punctuation left stranded by removing a filler.

    The filler pattern consumes a comma *after* the word but not before it, so a
    filler that closed a clause left the comma dangling — "this is probably fine,
    you know" became "this is probably fine," and "the tests, you know, are slow"
    became "the tests, are slow". Both are text typed into someone's document, and
    both read as a bug in the tool rather than as speech.

    Only ever applied when the filler pass actually removed something, so text
    nobody touched keeps whatever punctuation was dictated.
    """
    # ", ," -> ","  (filler sat between two commas)
    text = re.sub(r",\s*,", ",", text)
    # " ," -> ","   (the word before the comma was the one removed)
    text = re.sub(r"\s+([,;:])", r"\1", text)
    # a clause separator with nothing left after it
    text = re.sub(r"[,;:]\s*$", "", text)
    return text


def _remove_fillers(text: str, filler_words: list[str]) -> str:
    if not filler_words:
        return text

    # A hyphen is a word boundary, so Rule A would otherwise reach *inside* a
    # hyphenated token and delete one part of it — turning "right-click" into
    # "-click" and, worse, "a-a-actually" (what a stutter looks like once
    # transcribed) into a fragment glued to the next word. Since this filter is
    # the implementation of dysfluency-friendly mode, destroying a stuttered
    # word is the worst possible failure for the user it exists to serve.
    #
    # So hyphenated tokens are handled here, at token level, and never opened
    # up by the regex below: a token that is *entirely* filler is dropped whole,
    # and any other hyphenated token is left exactly as spoken. Collapsing
    # "a-a-actually" to "actually" would be the more ambitious repair, but it
    # cannot be done safely by a filler-removal pass — "w-w-want" proves the
    # repeated fragment is not always a filler — so that belongs to the
    # repetition rules (issue #144).
    fillers = frozenset(w.lower() for w in filler_words)
    text = _drop_all_filler_hyphen_tokens(text, fillers)
    # Sort longest first so multi-word fillers match before single words.
    # The trailing \b is load-bearing: without it "like" matched the prefix of
    # "likely" and left "ly" behind, and "actually" was eaten out of the middle
    # of a URL. Every filler is alphabetic, so a word boundary on both sides is
    # always the right anchor.
    sorted_fillers = sorted(filler_words, key=len, reverse=True)
    # The optional leading `,\s*` matters as much as the trailing one: a filler
    # is often a parenthetical ("the tests, you know, are slow"), and consuming
    # only the closing comma left "the tests, are slow" — a comma between subject
    # and verb, typed into the user's document.
    pattern = re.compile(
        r'(?:,\s*)?\b(' + '|'.join(re.escape(w) for w in sorted_fillers) + r')\b[,]?\s*',
        re.IGNORECASE,
    )

    def _enclosing_token(start: int, end: int) -> str:
        """The whitespace-delimited token(s) the match sits inside."""
        while start > 0 and not text[start - 1].isspace():
            start -= 1
        while end < len(text) and not text[end].isspace():
            end += 1
        return text[start:end]

    # Set when a filler is removed from position 0, so its own trailing
    # punctuation can be cleared afterwards ("Uh... so I think" must not be
    # left as "... so I think"). The regex cannot consume it inside the match
    # without also eating sentence punctuation after a mid-utterance filler.
    removed_leading = False

    def _replacer(m: re.Match) -> str:
        nonlocal removed_leading
        matched = m.group(0)
        # Protect uppercase / code tokens. This must test the *enclosing* token,
        # not the matched filler: "basically" inside "basically_fn" is a bare
        # lowercase word on its own and would sail past the guard, turning a code
        # identifier into "_fn" mid-dictation.
        enclosing = _enclosing_token(m.start(1), m.end(1))
        # Never reach inside a hyphenated token. Anything still containing a
        # hyphen at this point has at least one non-filler part, i.e. a real
        # word the user said; removing the filler part alone would corrupt it.
        if '-' in enclosing.strip(_TRAILING_PUNCTUATION):
            return matched
        at_start = _is_utterance_initial(text, m.start())
        if _is_protected(enclosing):
            # Exception (issue #117 / option 2, narrowed by #120 and #122):
            # Whisper capitalises the first word of an utterance, so
            # sentence-initial fillers ("Um …") trip the uppercase guard and
            # would otherwise never be removed. Relax the *uppercase* check only
            # at position 0, and only for fillers that cannot open a real
            # sentence.
            if (
                _is_code_or_path_token(enclosing)
                or not at_start
                or not _is_unambiguous_filler(m.group(1).lower())
            ):
                return matched
        if at_start:
            removed_leading = True
        # When the match swallowed the parenthetical's opening comma it also
        # swallowed the space in front of it, so returning nothing would glue the
        # neighbours together ("the tests, you know, are slow" -> "the testsare
        # slow"). One space; the collapse pass at the end tidies any doubling.
        return ' ' if matched.lstrip().startswith(',') else ''

    out = pattern.sub(_replacer, text)
    if removed_leading:
        out = out.lstrip(_TRAILING_PUNCTUATION + ' ')
    if out != text:
        out = _tidy_orphaned_punctuation(out)
    return out


def _dedup_2grams(text: str) -> str:
    """Remove second occurrence of consecutive identical 2-grams."""
    while True:
        tokens = text.split()
        changed = False
        i = 0
        result: list[str] = []
        while i < len(tokens):
            if (
                i + 3 < len(tokens)
                and tokens[i] == tokens[i + 2]
                and tokens[i + 1] == tokens[i + 3]
            ):
                result.append(tokens[i])
                result.append(tokens[i + 1])
                i += 4
                changed = True
            else:
                result.append(tokens[i])
                i += 1
        text = ' '.join(result)
        if not changed:
            break
    return text


# Words that, immediately before a self-correction trigger, mean the speaker is
# *talking about* the action rather than performing it. "Please do not delete that
# branch" is an instruction whose whole point is the negation; treating it as a
# rollback deleted the sentence and left the object of the instruction behind
# ("branch"), which is the most dangerous output the filter can produce — it does
# not garble the meaning, it inverts it.
_NEGATIONS_BEFORE_TRIGGER = (
    "not", "don't", "dont", "doesn't", "doesnt", "didn't", "didnt",
    "won't", "wont", "can't", "cant", "cannot", "never", "no",
)


def _trigger_is_negated(lower: str, idx: int) -> bool:
    """True when the trigger at *idx* is governed by a preceding negation.

    Only the word immediately before it is consulted. A wider window would start
    suppressing genuine corrections in long sentences that merely happen to
    contain a "not" somewhere earlier, and a missed rollback is a visible extra
    sentence while a wrong one is silently destroyed meaning.
    """
    before = lower[:idx].rstrip(" ,")
    if not before:
        return False
    return before.rsplit(" ", 1)[-1] in _NEGATIONS_BEFORE_TRIGGER


def _trigger_positions(lower: str, trigger: str):
    """Every position where *trigger* is a real rollback, earliest first.

    ``str.find`` matched anywhere, including inside ordinary prose, so a sentence
    that merely contains the words fired a rollback that discarded everything
    before it. Each candidate is filtered on word boundaries and on whether a
    negation governs it.
    """
    start = 0
    while (idx := lower.find(trigger, start)) != -1:
        start = idx + 1
        before_ok = idx == 0 or not (lower[idx - 1].isalnum() or lower[idx - 1] == "'")
        after = idx + len(trigger)
        after_ok = after >= len(lower) or not (lower[after].isalnum() or lower[after] == "'")
        if before_ok and after_ok and not _trigger_is_negated(lower, idx):
            yield idx


def _apply_self_corrections(text: str, triggers: list[str]) -> str:
    if not triggers:
        return text
    while True:
        lower = text.lower()
        # Resolve the trigger that was *spoken* first, not the one that happens
        # to be declared first in config (#145). Iterating the config list meant
        # "scratch that. no wait …" resolved "no wait" first purely because it
        # sits earlier in the default list, discarding the correction the user
        # actually made first.
        found = min(
            (
                (idx, trigger)
                for trigger in triggers
                for idx in _trigger_positions(lower, trigger.lower())
            ),
            default=None,
        )
        if found is None:
            return text
        idx, trigger = found

        # Find last sentence boundary before the trigger
        boundary = max(
            text.rfind('. ', 0, idx),
            text.rfind('! ', 0, idx),
            text.rfind('? ', 0, idx),
        )
        end = idx + len(trigger)
        # Consume the punctuation that belongs to the trigger clause. Periods
        # were previously left behind, so "delete that." — the ordinary way
        # people phrase it, as a complete sentence — deposited a bare " . " in
        # the middle of the output (#145).
        while end < len(text) and text[end] in ' ,.!?…':
            end += 1
        if boundary == -1:
            # No sentence boundary — remove everything up to end of trigger
            text = text[end:].strip()
        else:
            # Remove from after the boundary separator through end of trigger
            keep_end = boundary + 2  # keep '. '
            text = (text[:keep_end] + text[end:]).strip()

File: android/core/postprocess/src/main/kotlin/com/yazses/core/postprocess/Disfluency.kt
package com.yazses.core.postprocess

/**
 * Settings for [filterTranscript].
 *
 * Mirrors the desktop `[filters.disfluency]` section. Defaults match the desktop
 * defaults, which the contract vectors assert against when a case supplies no
 * options.
 */
public data class DisfluencyConfig(
    val enabled: Boolean = true,
    /**
     * Words removed as fillers.
     *
     * Deliberately short. `like`, `right`, `sort of`, `kind of`, `actually`,
     * `basically` and `literally` are absent: each is a genuine filler in some
     * positions and load-bearing content in others, and stripping them turns
     * hedges into facts ("it seems basically correct" is not "it seems correct").
     * Someone who wants aggressive removal adds them back.
     */
    val fillerWords: List<String> = listOf(
        "um", "uh", "er", "ah", "hmm", "you know",
        "i mean", "okay so",
        "so um", "so uh",
    ),
    val selfCorrectionTriggers: List<String> = listOf(
        "no wait", "delete that", "scratch that", "never mind",
        "forget that", "strike that",
    ),
    /** Dysfluency-Friendly Mode collapse pass; off by default. */
    val collapseRepetitions: Boolean = false,
    val collapseProlongations: Boolean = false,
    val prolongationMinRun: Int = 3,
    val repetitionMaxFragmentLen: Int = 2,
)

/** Result of the filter: the text, and how much shorter it got. */
public data class FilterResult(val text: String, val charsRemoved: Int)

private const val TRAILING_PUNCTUATION = ".,;:!?…"

/** Non-lexical hesitation sounds — never ordinary words, so safe at position 0. */
private val HESITATION_PARTICLES = setOf("um", "uh", "er", "ah", "hmm")

/**
 * Words that, immediately before a self-correction trigger, mean the speaker is
 * *talking about* the action rather than performing it.
 *
 * "Please do not delete that branch" is an instruction whose whole point is the
 * negation, and treating it as a rollback left only "branch" — the most dangerous
 * output this filter can produce, because it does not garble the meaning, it
 * inverts it. Modals do the same thing without a negation ("you should never mind
 * the warning"): they put the trigger inside a verb phrase, and a correction is an
 * interjection that never continues the clause it interrupts.
 */
private val GOVERNING_WORDS = setOf(
    "not", "don't", "dont", "doesn't", "doesnt", "didn't", "didnt",
    "won't", "wont", "can't", "cant", "cannot", "never", "no",
    "should", "would", "could", "can", "will", "shall", "must", "may", "might",
    "do", "does", "did", "to", "please", "let", "lets", "let's",
)

/**
 * The same evidence, for the two phrase types a *verb* guard cannot see.
 *
 * A determiner or possessive requires a noun head, so a trigger after one is part of
 * a noun phrase and cannot be an interjection: "the no wait policy applies to
 * walk-ins" became "policy applies to walk-ins". A reporting verb makes the trigger
 * *quoted* rather than performed — the speaker is describing a correction somebody
 * else made, so "he said never mind the cost and left" became "the cost and left". A
 * copula covers the residue where the trigger heads a predicate noun phrase and the
 * determiner sits inside the trigger itself, as in "there is no wait time at this
 * branch", where the word before "no wait" is "is".
 *
 * Demonstratives are deliberately absent: "this" and "that" are as often the *object*
 * of the preceding verb as a determiner on the trigger, and "please do this strike
 * that please do that" is a real correction a demonstrative entry would suppress.
 *
 * Kept as a second set rather than folded into [GOVERNING_WORDS] so each half stays
 * comparable to the Python constant it mirrors — the contract vectors name only one
 * of these 33 words, so vector agreement cannot be what keeps the two in step.
 */
private val PHRASE_CONTEXT_WORDS = setOf(
    // articles and possessives
    "the", "a", "an", "my", "your", "our", "their", "its", "his", "her",
    // reporting verbs — the trigger is quoted, not performed
    "said", "says", "say", "saying", "told", "tells", "tell", "asked", "asks",
    "wrote", "writes", "replied", "answered", "shouted", "yelled", "whispered",
    // copulas
    "is", "are", "was", "were", "be", "been", "being",
    // Nominative-only personal pronouns: a subject immediately before the trigger
    // makes the trigger the clause's own verb. "it" and "you" are deliberately
    // absent -- they are also objects, and a real correction turns on one of them
    // ("i think we should ship it never mind lets wait").
    "i", "we", "they", "he", "she",
)

private val DOUBLE_SPACE = Regex("""  +""")

private fun isProtected(token: String): Boolean =
    token.any { it.isUpperCase() } || '_' in token || '/' in token || '.' in token

private fun isCodeOrPathToken(token: String): Boolean {
    val core = token.trimEnd { it in TRAILING_PUNCTUATION }
    return '_' in core || '/' in core || '.' in core
}

private fun isUnambiguousFiller(filler: String): Boolean =
    filler.split(" ").any { it in HESITATION_PARTICLES }

/**
 * Three passes, then a self-correction rollback.
 *
 * Behaviour is defined by the disfluency contract vectors, not by this code. The
 * Kotlin is idiomatic rather than a transliteration of the Python (ADR-MOB-008 §4).
 */
public fun filterTranscript(text: String, config: DisfluencyConfig = DisfluencyConfig()): FilterResult {
    if (!config.enabled || text.isBlank()) return FilterResult(text, 0)

    val originalLength = text.length
    var out = removeFillers(text, config.fillerWords)
    out = dedup2Grams(out)
    out = collapseDysfluencies(out, config)
    out = applySelfCorrections(out, config.selfCorrectionTriggers)
    out = DOUBLE_SPACE.replace(out, " ").trim()

    return FilterResult(out, maxOf(0, originalLength - out.length))
}

// ---- Rule A: filler removal ------------------------------------------------

private fun dropAllFillerHyphenTokens(text: String, fillers: Set<String>): String {
    // A hyphen is a word boundary, so the regex below would otherwise reach
    // *inside* a hyphenated token and delete one part of it — turning
    // "a-a-actually" (what a stutter looks like once transcribed) into a fragment
    // glued to the next word. Destroying a stuttered word is the worst possible
    // failure for the user this filter exists to serve.
    val kept = text.split(" ").filter { token ->
        val core = token.trim { it in TRAILING_PUNCTUATION }
        if ('-' !in core) return@filter true
        val parts = core.split("-").filter { it.isNotEmpty() }
        !(parts.isNotEmpty() && parts.all { it.lowercase() in fillers })
    }
    return kept.joinToString(" ")
}

private fun removeFillers(text: String, fillerWords: List<String>): String {
    if (fillerWords.isEmpty()) return text
    val fillers = fillerWords.map { it.lowercase() }.toSet()
    val withoutHyphenFillers = dropAllFillerHyphenTokens(text, fillers)

    // Longest first so a multi-word filler matches before any single word could
    // claim part of it. The word boundaries are load-bearing: without the trailing
    // one, "like" matched the prefix of "likely" and left "ly" behind.
    //
    // The optional leading comma matters as much as the trailing one: a filler is
    // often a parenthetical ("the tests, you know, are slow"), and consuming only
    // the closing comma left "the tests, are slow" — a comma between subject and
    // verb, typed into the user's document.
    val alternation = fillerWords.sortedByDescending { it.length }.joinToString("|") { Regex.escape(it) }
    val pattern = Regex("""(?:,\s*)?\b($alternation)\b[,]?\s*""", RegexOption.IGNORE_CASE)

    var removedLeading = false
    val result = StringBuilder()
    var last = 0
    for (m in pattern.findAll(withoutHyphenFillers)) {
        val group = m.groups[1]!!.range
        val enclosing = enclosingToken(withoutHyphenFillers, group.first, group.last + 1)
        val atStart = withoutHyphenFillers.substring(0, m.range.first).isBlank()

        val keep = when {
            // Never reach inside a hyphenated token: anything still containing a
            // hyphen has at least one non-filler part, i.e. a real word.
            '-' in enclosing.trim { it in TRAILING_PUNCTUATION } -> true
            isProtected(enclosing) -> {
                // Whisper capitalises the first word, so a sentence-initial "Um"
                // trips the uppercase guard and would never be removed. Relax it
                // at position 0, and only for fillers that cannot open a sentence.
                isCodeOrPathToken(enclosing) || !atStart || !isUnambiguousFiller(m.groupValues[1].lowercase())
            }
            else -> false
        }

        result.append(withoutHyphenFillers, last, m.range.first)
        if (keep) {
            result.append(m.value)
        } else {
            if (atStart) removedLeading = true
            // When the match swallowed the parenthetical's opening comma it also
            // swallowed the space in front of it; returning nothing would glue the
            // neighbours together. The collapse at the end tidies any doubling.
            if (m.value.trimStart().startsWith(",")) result.append(' ')
        }
        last = m.range.last + 1
    }
    result.append(withoutHyphenFillers, last, withoutHyphenFillers.length)

    var out = result.toString()
    if (removedLeading) out = out.trimStart { it in TRAILING_PUNCTUATION || it == ' ' }
    if (out != withoutHyphenFillers) out = tidyOrphanedPunctuation(out)
    return out
}

private fun enclosingToken(text: String, startIn: Int, endIn: Int): String {
    var start = startIn
    var end = endIn
    while (start > 0 && !text[start - 1].isWhitespace()) start--
    while (end < text.length && !text[end].isWhitespace()) end++
    return text.substring(start, end)
}

/**
 * Repair punctuation left stranded by removing a filler.
 *
 * Only applied when the filler pass actually removed something, so text nobody
 * touched keeps whatever punctuation was dictated.
 */
private fun tidyOrphanedPunctuation(text: String): String =
    text
        .replace(Regex(""",\s*,"""), ",")
        .replace(Regex("""\s+([,;:])"""), "$1")
        .replace(Regex("""[,;:]\s*$"""), "")

// ---- Rule B: consecutive 2-gram dedup --------------------------------------

private fun dedup2Grams(text: String): String {
    var current = text
    while (true) {
        val tokens = current.split(" ").filter { it.isNotEmpty() }
        val result = mutableListOf<String>()
        var changed = false
        var i = 0
        while (i < tokens.size) {
            if (i + 3 < tokens.size &&
                tokens[i] == tokens[i + 2] &&
                tokens[i + 1] == tokens[i + 3]
            ) {
                result += tokens[i]
                result += tokens[i + 1]
                i += 4
                changed = true
            } else {
                result += tokens[i]
                i++
            }
        }
        current = result.joinToString(" ")
        if (!changed) return current
    }
}

// ---- Rule B.5: stutter / prolongation collapse (opt-in) --------------------

private fun collapseProlongations(text: String, minRun: Int): String {
    if (minRun < 2) return text
    val pattern = Regex("""([a-zA-Z])\1{${minRun - 1},}""")
    return text.split(" ").joinToString(" ") { token ->
        if (isProtected(token)) token else pattern.replace(token, "$1")
    }
}

private fun fixHyphenStutter(token: String): String {
    if ('-' !in token || isProtected(token)) return token
    val parts = token.split("-")
    if (parts.size < 3) return token
    val lead = parts.dropLast(1)
    val final = parts.last()
    val fragment = lead.first()
    val allSame = fragment.isNotEmpty() && fragment.all { it.isLetter() } && lead.all { it == fragment }
    return if (allSame && final.isNotEmpty() && final.lowercase().startsWith(fragment.lowercase())) final else token
}

private fun collapseRepetitions(text: String, maxFragmentLen: Int): String {
    val tokens = text.split(" ").filter { it.isNotEmpty() }.map(::fixHyphenStutter)
    val result = mutableListOf<String>()
    var i = 0
    while (i < tokens.size) {
        val token = tokens[i]
        // Two or more identical short fragments followed by the longer word they
        // prefix: "b b because" -> "because".
        if (token.isNotEmpty() && token.all { it.isLetter() } && !isProtected(token) && token.length <= maxFragmentLen) {
            var j = i
            while (j < tokens.size && tokens[j] == token) j++
            if (j - i >= 2 && j < tokens.size && !isProtected(tokens[j]) &&
                tokens[j].length > token.length &&
                tokens[j].lowercase().startsWith(token.lowercase())
            ) {
                result += tokens[j]
                i = j + 1
                continue
            }
        }
        // A run of three or more identical tokens collapses to one.
        if (!isProtected(token)) {
            var j = i
            while (j < tokens.size && tokens[j] == token) j++
            if (j - i >= 3) {
                result += token
                i = j
                continue
            }
        }
        result += token
        i++
    }
    return result.joinToString(" ")
}

private fun collapseDysfluencies(text: String, config: DisfluencyConfig): String {
    var out = text
    if (config.collapseRepetitions) out = collapseRepetitions(out, config.repetitionMaxFragmentLen)
    if (config.collapseProlongations) out = collapseProlongations(out, config.prolongationMinRun)
    return out
}

// ---- Rule C: self-correction rollback --------------------------------------

private fun triggerIsGoverned(lower: String, idx: Int): Boolean {
    val before = lower.substring(0, idx).trimEnd(' ', ',')
    if (before.isEmpty()) return false
    val preceding = before.substringAfterLast(' ')
    return preceding in GOVERNING_WORDS || preceding in PHRASE_CONTEXT_WORDS
}

/**
 * True when something precedes the trigger for it to correct.
 *
 * A self-correction replaces what was just said, so a trigger with nothing in
 * front of it cannot be one — it is a sentence starting with those words:
 * "delete that file when you are done", "never mind the warning". Each of those
 * previously discarded everything before the trigger and left the remainder,
 * which is the worst thing this filter can do: the result reads as fluent text
 * the user never said.
 *
 * At position 0 the only signal is the punctuation Whisper writes for the pause,
 * and the safe default is to require it. Deliberately NOT required mid-utterance:
 * Whisper does not reliably render a spoken pause, and corrections without one
 * are a supported case.
 */
private fun hasTextToRollBack(lower: String, start: Int, end: Int): Boolean {
    if (lower.substring(0, start).isNotBlank()) return true
    return lower.getOrNull(end) in setOf('.', ',', ';', ':', '!', '?', '…')
}

private fun triggerPositions(lower: String, trigger: String): List<Int> {
    val out = mutableListOf<Int>()
    var start = 0
    while (true) {
        val idx = lower.indexOf(trigger, start)
        if (idx < 0) return out
        start = idx + 1
        val beforeOk = idx == 0 || !(lower[idx - 1].isLetterOrDigit() || lower[idx - 1] == '\'')
        val after = idx + trigger.length
        val afterOk = after >= lower.length || !(lower[after].isLetterOrDigit() || lower[after] == '\'')
        if (beforeOk && afterOk && !triggerIsGoverned(lower, idx) &&
            hasTextToRollBack(lower, idx, after)
        ) {
            out += idx
        }
    }
}

private fun applySelfCorrections(text: String, triggers: List<String>): String {
    if (triggers.isEmpty()) return text
    var current = text
    while (true) {
        val lower = current.lowercase()
        // Resolve the trigger that was SPOKEN first, not the one declared first in
        // config: "scratch that. no wait ..." must resolve "scratch that".
        val found = triggers
            .flatMap { t -> triggerPositions(lower, t.lowercase()).map { it to t.lowercase() } }
            .minByOrNull { it.first } ?: return current

        val (idx, trigger) = found

        val boundary = maxOf(
            current.lastIndexOf(". ", idx - 1),
            current.lastIndexOf("! ", idx - 1),
            current.lastIndexOf("? ", idx - 1),
        )
        var end = idx + trigger.length
        // Consume the punctuation belonging to the trigger clause, or "delete
        // that." leaves a bare " . " in the middle of the output.
        while (end < current.length && current[end] in " ,.!?…") end++

        current = if (boundary == -1) {
            current.substring(end).trim()
        } else {
            (current.substring(0, boundary + 2) + current.substring(end)).trim()
        }
    }
}
